core/src/external/shell.rs

```diff
@@ -49,3 +50,166 @@ pub fn shell(opt: ShellOpt) -> Result<Child> {
+				iter.next();
+				let mut text = String::new();
+				loop {
+					if iter.peek().is_none() {
+						break;
+					}
+					if iter.peek().is_some_and(|_c| *_c == '"') {
+						iter.next();
+						break;
+					}
+					get_next_char(&mut iter, &mut text, args);
+				}
+				expanded_args.push(text);
+			} else {
+				if *c == '%' {
+					let mut tmp_iter = iter.clone();
+					tmp_iter.next();
+					if tmp_iter.peek().is_some_and(|_c| *_c == '*') {
+						iter.next();
+						iter.next();
+						for arg in args {
+							expanded_args.push(arg.as_ref().to_string())
+						}
+						continue;
+					}
+				}
+
+				let mut text = String::new();
+				loop {
+					if iter.peek().is_none() || iter.peek().is_some_and(|_c| _c.is_whitespace()) {
+						break;
+					}
+					get_next_char(&mut iter, &mut text, args);
+				}
+				expanded_args.push(text);
+			}
+		}
+
+		expanded_args
+	}
+
+	fn get_next_char<T>(
+		iter: &mut std::iter::Peekable<std::str::Chars<'_>>,
+		text: &mut String,
+		args: &[T],
+	) where
+		T: AsRef<str>,
+	{
+		let ch = iter.next().unwrap();
+		if ch == '\\' {
+			match iter.next() {
+				Some('n') => text.push('\n'),
+				Some('r') => text.push('\r'),
+				Some('t') => text.push('\t'),
+				Some(x) => text.push(x),
+				None => (),
+			}
+		} else if ch == '%' {
+			if iter.peek().is_some_and(|_c| *_c == '*') {
+				iter.next();
+				text.push_str(&args.iter().map(|value| value.as_ref()).collect::<Vec<&str>>().join(" "));
+			} else {
+				let mut num = String::new();
+				while iter.peek().is_some_and(|_c| _c.is_numeric()) {
+					num.push(iter.next().unwrap());
+				}
+				if num.is_empty() {
+					text.push('%');
+				} else {
+					let i: usize = num.parse().unwrap();
+					if i > 0 {
+						text.push_str(args.get(i - 1).map(|value| value.as_ref()).unwrap_or_default());
+					}
+				}
+			}
+		} else {
+			text.push(ch);
+		}
+	}
+
+	#[cfg(test)]
+	mod tests {
+		use super::*;
+
+		#[test]
+		fn test_no_quote() {
+			let args = parse_cmd_to_args("echo abc xyz %1 %2", &["111", "222"]);
+			assert_eq!(args, vec!["echo", "abc", "xyz", "111", "222"]);
+
+			let args = parse_cmd_to_args("  echo   abc   xyz %1   %2  ", &["111", "222"]);
+			assert_eq!(args, vec!["echo", "abc", "xyz", "111", "222"]);
+		}
+
+		#[test]
+		fn test_single_quote() {
+			let args = parse_cmd_to_args("echo 'abc xyz' '%1' %2", &["111", "222"]);
+			assert_eq!(args, vec!["echo", "abc xyz", "111", "222"]);
+
+			let args = parse_cmd_to_args("echo 'abc \"\"xyz' '%1' %2", &["111", "222"]);
+			assert_eq!(args, vec!["echo", "abc \"\"xyz", "111", "222"]);
+		}
+
+		#[test]
+		fn test_double_quote() {
+			let args = parse_cmd_to_args("echo \"abc ' 'xyz\" \"%1\" %2 %3", &["111", "222"]);
+			assert_eq!(args, vec!["echo", "abc ' 'xyz", "111", "222", ""]);
+		}
+
+		#[test]
+		fn test_escaped() {
+			let args = parse_cmd_to_args("echo \"a\tbc ' 'x\nyz\" \"\\%1\" %2 %3", &["111", "22  2"]);
+			assert_eq!(args, vec!["echo", "a\tbc ' 'x\nyz", "%1", "22  2", ""]);
```

Comment:
This test case seems wrong; the fourth element should be `"\\%1"` instead of `"%1"`?

Reply:
Got it!